### PR TITLE
Flux Batched Inference

### PR DESCRIPTION
**Pull request by wesleytruong**
This PR reuses the trainer's parallelization to perform batched inference on Flux. It follows up from and addresses comments on @CarlosGomes98 https://github.com/pytorch/torchtitan/pull/1227. This irons out some optimizations for better code clarity.

In this implementation saves images locally from each rank but maintains the global information such as prompt and idx of each image.  It also removes unnecessary padding and handles prompt sets that are not divisible by the batch size.

### CODE REVIEW

File: torchtitan/experiments/flux/inference/run_infer.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+# e.g.
+# LOG_RANK=0,1 NGPU=4 ./torchtitan/experiments/flux/run_train.sh
+NGPU=${NGPU:-"8"}
+export LOG_RANK=${LOG_RANK:-0}
+CONFIG_FILE=${CONFIG_FILE:-"./torchtitan/experiments/flux/train_configs/debug_model.toml"}
+
+overrides=""
+if [ $# -ne 0 ]; then
+    overrides="$*"
+fi
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/inference/run_infer.sh]: this can be improved, see `run_train.sh`

File: torchtitan/experiments/flux/sampling.py [hunk 2]
```diff
@@ -206,8 +219,15 @@ def denoise(
             pred_u, pred_c = pred.chunk(2)
             pred = pred_u + classifier_free_guidance_scale * (pred_c - pred_u)
 
+            # repeat along batch dimension to update both unconditional and conditional latents
+            pred = pred.repeat(2, 1, 1)
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/sampling.py:223]: wait what happens before and after for non-batched sampling?
Did we do repeat (unconditional + conditional) before?

File: torchtitan/experiments/flux/dataset/tokenizer.py [hunk 3]
```diff
@@ -45,13 +45,31 @@ def _pad_and_chunk_tokens(
     def get_vocab_size(self) -> int:
         return self.tiktokenizer.vocab_size
 
-    def encode(self, text: str) -> torch.Tensor:
+    def encode(self, text: str | list[str]) -> torch.Tensor:
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/tokenizer.py:48]: It seems you are doing batchify within tokenizer. Can't we rely on dataloader to do this for us, as what happens in training?

File: torchtitan/experiments/flux/inference/infer.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
+# Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/inference/infer.py]: NVIDIA??? lol

File: torchtitan/experiments/flux/inference/infer.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+from torchtitan.experiments.flux.dataset.tokenizer import (
+    build_flux_tokenizer,
+    FluxTokenizer,
+)
+from torchtitan.experiments.flux.sampling import generate_image, save_image
+from torchtitan.experiments.flux.train import FluxTrainer
+from torchtitan.tools.logging import init_logger, logger
+
+
+def inference_call(
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/inference/infer.py]: I feel this is too minimal to be a standalone function.

Also can do generate + save for each batch? Not sure why it helps to generate all and then save all.

File: torchtitan/experiments/flux/inference/infer.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        # Generate images for this process's assigned prompts
+        images = inference_call(
+            prompts,
+            trainer,
+            t5_tokenizer,
+            clip_tokenizer,
+            bs=config.inference.batch_size,
+        )
+
+        if config.inference.save_img_folder:
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/inference/infer.py]: why `if`?

File: torchtitan/experiments/flux/inference/infer.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                # Extract single image while preserving batch dimension [1, C, H, W]
+                img = images[i : i + 1]
+                global_id = global_ids[i]
+
+                save_image(
+                    name=f"image_prompt{global_id}_rank{str(torch.distributed.get_rank())}.png",
+                    output_dir=os.path.join(
+                        config.job.dump_folder,
+                        config.inference.save_img_folder,
+                    ),
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/inference/infer.py]: only need to do this once outside the for-loop

File: torchtitan/experiments/flux/sampling.py (hunk 2 above)
Comment (wesleytruong) [reply torchtitan/experiments/flux/sampling.py:223]: The repeat for `pred` was previously getting handled automatically by broadcast here `latents = latents + (t_prev - t_curr) * pred`.

Previously, the `latents` would always be batch size 1 or 2 depending on if it is using cfg, and `pred` is always batch size 1. Since the new implementation accepts batch sizes of greater than 1, `latents` and `pred` won't be broadcast automatically in the case of cfg since `latents` will have twice the batch size of `pred` after `pred_u` and `pred_c` are combined , and broadcast only works if `pred`'s batch size is 1.

File: torchtitan/experiments/flux/dataset/tokenizer.py (hunk 3 above)
Comment (wesleytruong) [reply torchtitan/experiments/flux/dataset/tokenizer.py:48]: Here we are not splitting up the data into batches, but rather accepting a batch (list of strings) and encoding the entire thing to return as tensor. I did this since this way matches the behavior of the real FluxTokenizer and I think it's also more convenient to do this than to encode every string in a batch and then convert the returned lists of ids into tensor format in the program that calls `encode`

Comment (tianyu-l) [reply torchtitan/experiments/flux/dataset/tokenizer.py:48]: great, sounds good to me.

**Review by tianyu-l — APPROVED** [review]
Looks solid!
